### PR TITLE
pfblockerng: stop capturing hosts-dialect '##' comments as ABP rows

### LINKED ISSUE #1276: DNSBL: '##'-led whole-line comments (hosts-dialect section headers) are captured as ABP rows
Found by the post-merge audit of the 2026-07-10..07-13 devel batch (5963508a..603845d2).

## The bug

`pfb_dnsbl_is_abp_rule_line()`'s element-hiding branch returns TRUE for `$first === 0` (`pfblockerng.inc:7388-7391`) — a line **starting** with `##` is treated as a generic cosmetic rule. That is correct ABP syntax, but hosts-dialect feeds routinely use `## Section header` whole-line comments, and those now:

- are captured verbatim as ABP rows (`:17292-17296`) and staged as NDJSON `abp` rows every build;
- inflate `$alias_cnt` / widget stats (`grep -c ^` counts them, `:17155`, `:17581`);
- get re-parsed and dropped by Python on every load (`parse_abp` skips element-hiding rules, `pfb_unbound.py:4025`).

Pre-ADR-62, any `#`-led line hit the `#`-classifier's `continue` and was never staged or counted.

Probe (executed): `pfb_dnsbl_is_abp_rule_line('## hosts-style comment')` returns TRUE, while the docblock's own hostile analysis (`' ## '` mid-line, `'# c ## d'`) correctly stays FALSE.

## Impact

No blocking-behaviour change — junk rows, count inflation, and per-build parse waste.

## Fix shape

The `$first === 0` case needs to distinguish a real generic cosmetic rule (`##selector`) from a hosts-style comment (`## text`); e.g. require a non-empty selector immediately after the marker (no space). Keep the PHP and Python predicates in lockstep (`_dnsbl_is_abp_rule_line`).

---

🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

### PR DESCRIPTION
**Pull request by andrebrait**
## Summary

`pfb_dnsbl_is_abp_rule_line()` (PHP) and its Python mirror `_dnsbl_is_abp_rule_line()` wrongly captured a hosts-dialect whole-line comment like `## Section header` (a `##`/`#@#`/`#?#`/`#%#`/`#$#` marker at line position 0, followed by whitespace) as a real ABP element-hiding cosmetic rule — inflating `$alias_cnt`/widget stats and getting re-parsed and dropped by Python on every build. A real generic cosmetic rule (`##selector`, no space after the marker) must still classify TRUE, and does.

Fix: a marker at position 0 now also requires a non-whitespace, non-empty character immediately after it before the capture guard claims the line. The domain-prefixed branch (marker not at position 0) is untouched. Applied in lockstep to both the PHP predicate and its Python mirror, per ADR-62.

Fixes #1276

## Test plan

- [x] Red: `tests/php/Adr62DnsblIsAbpRuleLineTest.php` — 5 new DataProvider rows fail pre-fix (4 false-expecting rows return true)
- [x] Red: `tests/test_adr62_parity_oracle.py::test_cosmetic_prefix_guard_matches_php_row_for_row` — new rows fail pre-fix
- [x] Green: same tests, zero edits, pass post-fix
- [x] `vendor/bin/phpunit` full suite: 3355 tests, 0 failures
- [x] `composer phpstan`: no errors
- [x] `composer phpcs -- --standard=phpcs.xml.dist src/`: clean
- [x] `python3 -m pytest` full suite: 2876 passed, 4 skipped
- [x] `ruff check .` / `ruff format --check .`: clean
- [x] `mypy tests/`: no issues
- [x] Independent Sonnet verifier re-ran every gate + the red→green proof via `scripts/agent/verify-red-proof.sh` — PASS

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Gr4Sj3PTxg6CCSEMo2KtNk

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved detection of ad-blocking cosmetic rules to avoid incorrectly classifying hosts-dialect `## ...` whole-line comments as rules when the marker is at the start of a line.
  - Markers at line start are now treated as rules only when the character immediately after the marker is not whitespace (including correct handling for NBSP); preserved prior validation for non-start markers.

- **Tests**
  - Expanded ABP-rule-shape and parity-oracle coverage with additional non-capturable/capturable cases, including whitespace edge cases and sibling marker variants (`#@#`, `#?#`, `#%#`, `#$#`).

### CODE REVIEW

File: src/usr/local/pkg/pfblockerng/pfblockerng.inc [hunk 1]
```diff
@@ -7696,15 +7696,26 @@ function pfb_dnsbl_is_abp_rule_line(string $line): bool {
...
-			return TRUE;
+		if ($first !== FALSE) {
+			if ($first === 0) {
+				// issue #1276: a hosts-dialect '## Section' comment shares the
+				// marker-at-position-0 shape with a real cosmetic rule -- require
+				// a non-whitespace char immediately after the marker.
+				$after = substr($line, $marker_len, 1);
+				if ($after !== '' && $after !== ' ' && $after !== "\t") {
+					return TRUE;
+				}
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfblockerng.inc:7717]: Confirmed and applied — good catch. Executed probe against the code as it stood in this commit: `pfb_dnsbl_is_abp_rule_line("##\rSection")` (and `\x0B`/`\x0C` variants) returned `true`, since the guard only excluded literal `' '`/`"\t"`.

Fixed in 87e4fc9e: `$after !== '' && $after !== ' ' && $after !== "\t"` → `$after !== '' && !ctype_space($after)` — `ctype_space()` catches CR/VT/FF too while still correctly treating a bare marker (`$after === ''`) as no-selector. New hostile-input rows (marker+CR, marker+VT, marker+FF) added test-first to `Adr62DnsblIsAbpRuleLineTest.php` — red before the fix, green after, zero test edits between.

Note: `ctype_space()` is byte-wise and won't catch a multi-byte UTF-8 NBSP (`\xC2\xA0`) — this function is byte-oriented throughout (the domain-prefix `strspn` check has the same limitation), so full Unicode-whitespace parity with Python's `.isspace()` is a pre-existing architectural characteristic, not something narrowed by this fix.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

File: src/usr/local/pkg/pfblockerng/pfb_unbound.py [hunk 2]
```diff
@@ -3981,10 +3985,15 @@ def _dnsbl_is_abp_rule_line(line: str) -> bool:
...
-            first = min(positions)
-            if first == 0 or all(c in _DNSBL_ABP_COSMETIC_PREFIX_CHARS for c in line[:first]):
+        hits = [(line.find(m), len(m)) for m in _DNSBL_ELEMENT_HIDING_MARKERS]
+        hits = [(p, marker_len) for p, marker_len in hits if p != -1]
+        if hits:
+            first, marker_len = min(hits)
+            if first == 0:
+                after = line[marker_len : marker_len + 1]
+                if after and after not in (" ", "\t"):
+                    return True
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfb_unbound.py:3995]: Confirmed and applied. Executed probe: `_dnsbl_is_abp_rule_line("##\rSection")` (and `\x0b`/`\x0c`/NBSP variants) returned `True` pre-fix, since the guard only excluded literal `" "`/`"\t"`.

Fixed in 87e4fc9e: `after not in (" ", "\t")` → `not after.isspace()`, matching your suggestion exactly (Python's `.isspace()` on a decoded str also correctly treats a NBSP `"\xa0"` as whitespace — verified: `"\xa0".isspace()` → `True`). PHP now uses `ctype_space()` as the same-shaped native predicate, keeping both sides using "a whitespace predicate" per your note — PHP's is byte-wise so it won't catch a multi-byte UTF-8 NBSP the way Python's char-wise check does, but that's a pre-existing byte-vs-codepoint characteristic of this function (unrelated call sites like the domain-prefix check have the same limitation), not something this fix narrows further.

New hostile-input rows (marker+CR, marker+VT, marker+FF) added test-first to `test_adr62_parity_oracle.py` — red before the fix, green after, zero test edits between.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

**Comment by andrebrait** [comment]
## Review audit — PR #1343 (issue #1276)

**Reviews received:** CodeRabbit (clean), GitHub Copilot (2 findings), Claude adversarial review — dual-fallback (fable was rate-limited; ran one `sonnet` pass + one `opus` pass over the whole PR, per CLAUDE.md's top-tier-unavailable rule, findings unioned) plus 3 delta-scoped `sonnet` re-reviews (one per review-fix commit).

### CodeRabbit

"No actionable comments were generated in the recent review. 🎉" — clean pass, nothing to triage.

### GitHub Copilot

| # | Finding | File:line | Verdict |
|---|---|---|---|
| 1 | Position-0 marker guard only excluded literal `' '`/`"\t"` — missed CR/VT/FF | `pfblockerng.inc:7717` | **APPLIED** — `87e4fc9e` |
| 2 | Same gap, Python side; suggested `.isspace()` | `pfb_unbound.py:3995` | **APPLIED** — `87e4fc9e` |

Both threads replied with the fix commit + executed probes.

### Claude adversarial review (sonnet + opus whole-PR dual pass)

| # | Severity | Finding | Verdict |
|---|---|---|---|
| 1 | blocking (sonnet) / nitpick (opus) | Position-0 guard test coverage only exercised 2 of 5 element-hiding markers (`##`, `#@#`) — `#?#`/`#%#`/`#$#` untested | **APPLIED** — `516203e8` (6 new rows/language) |
| 2 | blocking (sonnet) | `ctype_space()` (PHP) vs `.isspace()` (Python) diverge on UTF-8 NBSP (`\xC2\xA0`) — a real lockstep break the widening commit introduced | **APPLIED** — `516203e8` (explicit NBSP byte-sequence check added to PHP) |

### Delta re-review 1 (of `87e4fc9e`, the CR/VT/FF widening)

| # | Severity | Finding | Verdict |
|---|---|---|---|
| F1 | blocking | Test-file comments cited "PR #1343 review:" — review archaeology per CLAUDE.md | **APPLIED** — `eecb365d` (reworded to plain `issue #1276:` refs; independently re-grepped the tree — the "15+ sibling files" precedent claimed by the whole-PR sonnet pass didn't hold, only 1 weak precedent existed) |
| F2 | blocking | PHP/Python NBSP divergence (same as above) | Already fixed in `516203e8` before this review landed — confirmed still holds |
| F3 | nitpick | `ctype_space()` is locale-dependent for bytes ≥0x80 — confirmed via executed probe: `ctype_space(chr(0xA0))` is `true` under this box's default `C.UTF-8` locale, `false` under `C` | **APPLIED** — `eecb365d` (swapped to a deterministic explicit `strspn()` byte set, matching the same function's own domain-prefix branch's existing convention; re-verified all ~70 test rows agree across `C`/`C.UTF-8`/`en_US.UTF-8`) |

### Delta re-review 2 (of `516203e8`, the NBSP + marker-coverage fix)

**PASS**, no blocking findings. 2 nitpicks:
- Residual Unicode-whitespace-class gap beyond NBSP (EM SPACE, IDEOGRAPHIC SPACE, etc. still diverge PHP vs Python) — confirmed real via executed probe, explicitly pre-existing/not worsened by this PR, deliberately scoped out (NBSP was the one realistic copy-paste artifact the issue named). **DEFERRED** — tracking issue #1345.
- Short-circuit evaluation order micro-nitpick — **SKIPPED**, reviewer's own analysis says negligible/not worth a fix.

### Delta re-review 3 (of `eecb365d`, the determinism + comment cleanup — final round)

**PASS**, no blocking findings. 2 informational-only notes on comment phrasing readability — not actionable, no fix needed.

### Gates (re-run at every commit, final state green)

- `vendor/bin/phpunit` (full): 3367 tests, 0 failures
- `composer phpstan`: no errors
- `composer phpcs -- --standard=phpcs.xml.dist src/`: clean
- `python3 -m pytest` (full): 2876 passed, 4 skipped
- `ruff check .` / `ruff format --check .`: clean
- `mypy tests/`: no issues
- CI (head `eecb365d`): all required checks green

### Findings ledger (complete)

1. CR/VT/FF whitespace gap (Copilot ×2) — fixed@87e4fc9e
2. Marker-family coverage gap (Claude sonnet+opus) — fixed@516203e8
3. NBSP PHP/Python parity break (Claude sonnet) — fixed@516203e8
4. Review-archaeology comments (delta review 1) — fixed@eecb365d
5. `ctype_space()` locale-dependence (delta review 1) — fixed@eecb365d
6. Residual non-NBSP Unicode whitespace gap (delta review 2) — deferred@#1345
7. Short-circuit order (delta review 2) — skipped, non-issue
8. Comment readability wording (delta review 3) — informational, no action

Every item has an outcome. Nothing left needing a human decision.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.
